### ISSUE TITLE
`workspace.onDidOpenTextDocument` gives stale, incorrectly located files

### ISSUE THREAD

**Issue by mrakgr** [microsoft/vscode#101286]
Version: 1.46.1 (user setup)
Commit: cd9ea6488829f560dc949a8b2fb789f3cdc05f5d
Date: 2020-06-17T21:13:20.174Z
Electron: 7.3.1
Chrome: 78.0.3904.130
Node.js: 12.8.1
V8: 7.8.279.23-electron.0
OS: Windows_NT x64 10.0.18363

Steps to Reproduce:

I have a file `package.spiproj` which looks like this:

```
dirOut: qwe
dirSource: c:\qwe
version: 1.2.3 (alpha)
name: Test
files: 
    A
    Submodule/
        QQ
        QQ
        WWWW
        E 
        Main
    Main
```

However the document in the `onDidOpenTextDocument` event is stale and shows up like this in the information window.

```
dirOut: qwe
dirSource: c:\qwe
version: 1.2.3 (alpha)
name: Test
files: 
    A
    Submodule/
        QQ 
        Q
        W
        E 
        Main
    Main
```

To make matters worse, its file name (in the event) is `package.spiproj.git` rather than `package.spiproj` which indicates that it is fetching it from cache somewhere. This is certainly the case as I do not have `package.spiproj.git` anywhere on my hard drive.

Right now I am doing a VS Code extension for the configuration files for my language and the way `workspace.onDidOpenTextDocument` behaves is both broken and useless for me. Even if I close the file and open it again, that event will still give me incorrect data. Also, I tried this using the Insiders build and am getting the same problem.

**Comment by jrieken** [comment]
Documents are addressed by URI which means you they have scheme, authority, path, query, and fragment (just like a web address). Do not assume that all document use the `file` scheme, there are many other schemes used and you must always check schemes

**Comment by mrakgr** [comment]
Even if I check the scheme (which is `git:/` in this case,) the problem is that it is fetching the stale version of the file from version control. If you look at the documentation for this function it states that...

```ts
/**
 * An event that is emitted when a [text document](#TextDocument) is opened or when the language id
 * of a text document [has been changed](#languages.setTextDocumentLanguage).
 *
 * To add an event listener when a visible text document is opened, use the [TextEditor](#TextEditor) events in the
 * [window](#window) namespace. Note that:
 *
 * - The event is emitted before the [document](#TextDocument) is updated in the
 * [active text editor](#window.activeTextEditor)
 * - When a [text document](#TextDocument) is already open (e.g.: open in another [visible text editor](#window.visibleTextEditors)) this event is not emitted
 *
 */
export const onDidOpenTextDocument: Event<TextDocument>;
```

More specifically it says that `The event is emitted before the document is updated in the active text editor`. Am I supposed to parse the uri to extract the actual file system path of the document and load the file separately? That seems very unlikely. There should be some function or event to get the text of the actual document being opened directly.

It is not good to me if the text of the document in that event is from a commit a few hours ago instead of the thing that will show up in the editor. So far I've managed to get along by using the `onDidChangeTextDocument` event which works as one would expect, but in order to synchronize the editor with the language server I do need to differentiate changes from initialization.

**Comment by jrieken** [comment]
Your assumption that `git:///some/path/some/file.txt.git` is the current file, as saved on disk or visible in the editor, is wrong. It's just a file, the git extension uses this as the baseline for diff computation but unless you truly understand what a scheme is used for you should simply ignore those documents

**Comment by mrakgr** [comment]
So you are saying that the Git extension is doing something under the hood to trigger that event? And that this event has nothing to do with the document actually being opened in the editor, is that right?

I see. Now that I am trying out opening various files after the plugin has activated, I see that the open event gets triggered twice for every file - once for the actual file open, and once by the git extension. The reason why the first one does not get triggered for the currently active file when the editor opens is probably because the extension I am working on is not active at that point. So instead of waiting for the event, I should probably look at the currently open files during activation of the extension...

Thank you. I see what I need to do now.

**Comment by jrieken** [comment]
>  And that this event has nothing to do with the document actually being opened in the editor, is that right?

Yeah. There is visibleEditors to know what documents are showing in what editor
